# Déjame escuchar la música

## Contenido:<a class="anchor" id="back"></a>

1. [Introducción](#introduccion)
2. [Etapa 1. Descripción de los datos](#data_review)
3. [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
4. [Etapa 3. Prueba de hipótesiss](#hypothesis)
8. [Conclusión](#conclusion)

## 1. Introducción <a class="anchor" id="introduccion"></a>

### Introducción <a id='introduccion'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








### Etapa 1. Descripción de los datos <a id='data_review'></a>

In [1]:
import os
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Nombre del archivo
archivo_nombre = 'music_project_en.csv'

# Rutas posibles para no Windows y Windows
url_no_windows = '/datasets/' + archivo_nombre
url_windows = 'datasets/' + archivo_nombre

# Verificar si el archivo existe en la ruta proporcionada
def cargar_archivo():
    """Intenta cargar el archivo desde diferentes rutas y manejar errores."""
    try:
        # Intentar cargar desde la ruta no-Windows
        if os.path.exists(url_no_windows):
            df = pd.read_csv(url_no_windows, sep=',', dtype={'votes': 'Int64'})
            print(f"Archivo cargado exitosamente desde '{url_no_windows}'")
            return df
        
        # Intentar cargar desde la ruta Windows
        elif os.path.exists(url_windows):
            df = pd.read_csv(url_windows, sep=',', dtype={'votes': 'Int64'})
            print(f"Archivo cargado exitosamente desde '{url_windows}'")
            return df
        
        else:
            raise FileNotFoundError(f"Error: el archivo '{archivo_nombre}' no se encuentra en ninguna de las rutas especificadas.")
    
    except FileNotFoundError as e:
        print(e)
    
    except pd.errors.ParserError as e:
        print(f"Error al procesar el archivo CSV: {e}")
    
    except Exception as e:
        print(f"Ha ocurrido un error inesperado: {e}")

# Llamar a la función de cargar archivo
df = cargar_archivo()

Archivo cargado exitosamente desde 'datasets/music_project_en.csv'


Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
#df = pd.read_csv('/datasets/music_project_en.csv')
# Obtener la información general sobre nuestros datos
print(list(df.columns))
 

['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day']


In [5]:
# Obtener las 10 primeras filas de la tabla df
display(df)

userID                              Track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina               NaN   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre       City        time        Day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
65074         rnb  Springfield  13:32:28  Wednesday  
65075         hip  Shelbyville  10:00:00     Monday  
65076  industrial  Springfield  20:09:26     Friday  
65077        rock  Springfield  21:43:59     Friday  
65078     country  Springfield  21:59:46     Friday  

[65079 rows x 7 columns]

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. hay que separa el nombre de la columna userId en formato tipo snake `Detecta el tercer problema por tu cuenta y descríbelo aquí`.




### Observaciones <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas` 
<br>&emsp;Los tipos de datos que se aprecian son del tipo String, del tipo Time y algunos datos Nan, se entiende la relacionacion que guarda el titulo de la columna con el contenido de las celdas, o por lo menos se espera obtener ese tipo de  informacion.


`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`
<br>&emsp;Si hay suficientes datos para dar respuesta a la hipotesis

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
<br>&emsp;Existen valores ausentes
**Correcion:**
Existen valores ausentes en las columnas *'track'*, *'artist'* y *'genre'*


## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [6]:
# Muestra los nombres de las columnas
print(list(df.columns))

['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day']


In [7]:

# Bucle en los encabezados poniendo todo en minúsculas
def df_to_minusculas(df):
    try: 
        tmp_df = df
        columns_new = []
        for columna in tmp_df.columns:
            columns_new.append(columna.lower()) 
        tmp_df.columns = columns_new
        return tmp_df
    except Exception as err: 
        print("------------------")
        print(f"Unexpected {err=}, {type(err)=}")
        print("------------------")
        return NameError
    
df_to_minusculas(df)
print(list(df.columns))

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


In [8]:
# Bucle para eliminar los espacios
def format_df(df):
    try: 
        tmp_df = df
        columns_new = []
        for columna in tmp_df.columns:
             columns_new.append(columna.lower().strip().replace(' ', '_'))
        tmp_df.columns = columns_new
        return tmp_df
    except Exception as err: 
        print("------------------")
        print(f"Unexpected {err=}, {type(err)=}")
        print("------------------")
        return NameError
    
format_df(df)
print(list(df.columns))

['userid', 'track', 'artist', 'genre', 'city', 'time', 'day']


In [9]:
format_df(df)
# Cambiar el nombre de la columna "userid"
df = df.rename(columns = {'userid': "user_id"})
print(list(df.columns))

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


In [10]:
# Comprobar el resultado: la lista de encabezados

print(list(df.columns))

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']


### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [11]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

In [12]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track', 'artist', 'genre']
for col in columns_to_replace: 
    df[col].fillna('unknown', inplace=True)
#print(df)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    65079 non-null  object
 2   artist   65079 non-null  object
 3   genre    65079 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [13]:
# Contar duplicados explícitos
print(df.duplicated().sum())

3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Eliminar duplicados explícitos
df = df.drop_duplicates()

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprobar de nuevo si hay duplicados
print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspeccionar los nombres de géneros únicos
x = df['genre'].sort_values(ascending=True).unique()
#ordenar

print(x) 

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [17]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre): # pasar una lista de valores incorrectos y un string con el valor correcto en la entrada de la función
    try:
        wrong_genres_len = int(df['genre'].nunique())
        for wrong_genre in wrong_genres: # un bucle sobre nombres mal escritos
            df['genre'] = df['genre'].replace(wrong_genre, correct_genre) # llamar a replace() para cada nombre incorrecto
        correct_genres_len = int(df['genre'].nunique())
        genres_modified = wrong_genres_len - correct_genres_len
        print("------------------")
        print(f"Total de filas afectadas = {genres_modified}")
        print("------------------")
        data = {'afected_rows':genres_modified,'initial_rows':wrong_genres_len,'final_rows':correct_genres_len}
        return data # devolver el DataFrame modificado
    except Exception as err: 
        print("------------------")
        print(f"Unexpected {err=}, {type(err)=}")
        print("------------------")
        return NameError
    


Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
# Eliminar duplicados implícitos
duplicados = ['hip', 'hop','hip-hop'] # una lista de nombres mal escritos
nombre_correcto = 'hiphop' # el nombre correcto
data = replace_wrong_genres(duplicados, nombre_correcto)
print (data)



------------------
Total de filas afectadas = 3
------------------
{'afected_rows': 3, 'initial_rows': 269, 'final_rows': 266}


Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos
print(df['genre'].sort_values(ascending=True).unique())
print (data)


['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

### Observaciones


<ins>*Antes de iniciar un calculo con los datos, necesitan darse formato, limpieza, evitar duplicados, valores nulos o vacios.
para eso se necesita que los datos sean tratados con unas funcionles que estan incluidas en Pandas, que nos permiten dejar los datos listos para su uso, como en el emplo anterior, de manera mas automatizada se pueden cargar y limpiar los datos desde una funcion previamente definida.*</ins>

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


In [20]:
# Contar las canciones reproducidas en cada ciudad 
genre_groups = df.groupby(by='city').count()
#print(genre_groups) #modificacion o anexo
print('--------------------------------------------------------------------------')
canciones_reproducidas = df.groupby(by='city')['track'].count()
print(canciones_reproducidas)


--------------------------------------------------------------------------
city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [21]:
# Calcular las canciones reproducidas en cada uno de los tres días
canciones_reproducidas = df.groupby(by='day')['track'].count()
print(canciones_reproducidas)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


`Observaciones`<br>
<ins>*Al parecer los dia de inicio de semana (lunes y viernes), son los dias que mas musica se escucha, ademas el viernes, se esucha todavia mas que el lunes*</ins>

In [22]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

def number_tracks(day=0,city=0):
    if (day == 0 or city == 0) or (  (type(day) is str) == False or (type(city) is str) == False ):
        print('Error al ingresar los datos')
        return None
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filas_por_dia =( df.loc[df['day'] == day] ) 
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    filas_por_columna  = filas_por_dia[filas_por_dia['city']== city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    usuarios = filas_por_columna['user_id'].count()
    # Devolve el número de valores de la columna 'user_id'
    return usuarios

print(number_tracks('Friday','Springfield'))

15945


In [23]:
# El número de canciones reproducidas en Springfield el lunes
print(number_tracks('Monday','Springfield'))

15740


In [24]:
# El número de canciones reproducidas en Shelbyville el lunes
print(number_tracks('Monday','Shelbyville'))

5614


In [25]:
# El número de canciones reproducidas en Springfield el miércoles
print(number_tracks('Wednesday','Springfield'))

11056


In [26]:
# El número de canciones reproducidas en Shelbyville el miércoles
print(number_tracks('Wednesday','Shelbyville'))

7003


In [27]:
# El número de canciones reproducidas en Springfield el viernes
print(number_tracks('Friday','Springfield'))

15945


In [28]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks('Friday','Shelbyville'))

5895


**Conclusion**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

> ### Objetivo:
>Prueba la hipótesis:
>1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

***
___De acuerdo a los datos obtenidos. `la Hipótesis es correcta`, dado que en cada uno de los dias de la semana los valores  distancian mucho unos de otros, ademas que tambien se ven reflejados en la ciudad en donde se escuchan___

## Conclusiones <a id='conclusion'></a>


***
___Se puede tomar como conclusion que en ambas ciudades y dependiendo el dia de la semana, hay una variacion considerable en los usuarios y las ususarias que escuchan musica en cierto dia de la semana.___
<br><br>
___Donde:___
<br>
- El dia *Viernes* en *Springfield* es el **mayor** dia donde se escucha música  en ambas ciudades.
- El dia *Lunes* en *Shelbyville* es el **menor** dia donde se escucha música  en ambas ciudades.
- En *Springfield* (42741 canciones/dia) se escucha ≈2.30% mas música que en *Shelbyville* (18512 canciones/dia). 
- El dia *viernes*  es el dia que se escucha más música en ambas ciudades con un valor de *21840* canciones/dia

**$\therefore$ la hipotesis propuesta el correcta.**
***
<br>

[Volver a la tabla de Contenidos](#back)